pkg/provisioner/filestorage/filestorage.go

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2018, Oracle and/or its affiliates. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filestorage
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/oracle/oci-go-sdk/common"
+	"github.com/oracle/oci-go-sdk/core"
+	"github.com/oracle/oci-go-sdk/filestorage"
+	"github.com/oracle/oci-go-sdk/identity"
+
+	"github.com/oracle/oci-volume-provisioner/pkg/oci/client"
+	"github.com/oracle/oci-volume-provisioner/pkg/provisioner/plugin"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ociVolumeID            = "volume.beta.kubernetes.io/oci-volume-id"
+	ociExportID            = "volume.beta.kubernetes.io/oci-export-id"
+	volumePrefixEnvVarName = "OCI_VOLUME_NAME_PREFIX"
+	fsType                 = "fsType"
+	subnetID               = "subnetId"
+	mntTargetID            = "mntTargetId"
+)
+
+// filesystemProvisioner is the internal provisioner for OCI filesystem volumes
+type filesystemProvisioner struct {
+	client client.ProvisionerClient
+}
+
+var _ plugin.ProvisionerPlugin = &filesystemProvisioner{}
+
+// NewFilesystemProvisioner creates a new file system provisioner that creates
+// filesystems using OCI File System Service.
+func NewFilesystemProvisioner(client client.ProvisionerClient) plugin.ProvisionerPlugin {
+	return &filesystemProvisioner{
+		client: client,
+	}
+}
+
+// getMountTargetFromID retrieves mountTarget from given mountTargetID
+func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, mountTargetID string) (*filestorage.MountTarget, error) {
+	resp, err := fsp.client.FileStorage().GetMountTarget(ctx, filestorage.GetMountTargetRequest{
+		MountTargetId: &mountTargetID,
+	})
+	if err != nil {
+		glog.Errorf("Failed to retrieve mount point mountTargetId=%q: %v", mountTargetID, err)
+		return nil, err
+	}
+	return &resp.MountTarget, nil
+}
+
+func (fsp *filesystemProvisioner) getOrCreateMountTarget(ctx context.Context, ad *string, subnetID string, compartmentID string) (*filestorage.MountTarget, error) {
+	// Check if there already is a mount target in the existing compartment
+	glog.Infof("Looking up existing mount targets")
+	responseListMnt, err := fsp.client.FileStorage().ListMountTargets(ctx, filestorage.ListMountTargetsRequest{
```

Comment:
Whilst the default limit is two mount targets the endpoint supports pagination so for future proofing we need to handle pagination.